### PR TITLE
Default Bundler Contributor Notes

### PR DESCRIPTION
**Pull request by AGawrys**
# ↪️ Pull Request

To read the docs formatted =>  https://github.com/parcel-bundler/parcel/blob/bundler-contributor-notes/docs/DefaultBundler.md

This is meant to be a step-by-step explanation of the bundler, specifically `createIdealGraph()` within the `DefaultBundler.js`. It's meant to be low-level-ish and reference variables, structures, and constraints in terms of logic. 

This should aid any future endeavors in creating a bundler, phasing out old structures, or supporting new features.

The PR contains visual aids for step-by-step examples as well as a document on manual bundling. 

## Before Merging
- [x] Merge in  [css-fix](https://github.com/parcel-bundler/parcel/pull/9023)
- [x] Merge in [manual bundling usage PR](https://github.com/parcel-bundler/website/pull/1099)
- [x] Address any comments

Once we merge in the source bundles PR, I will update all documents to reflect that change.

### CODE REVIEW

File: docs/DefaultBundler.md [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+
+We also create bundles for type change assets, parallel dependencies, and inline assets.
+
+```js
+parentAsset.type !== childAsset.type ||
+  dependency.priority === 'parallel' ||
+  childAsset.bundleBehavior === 'inline';
+```
+
+Bundles created by this path are **not** bundleGroups, and can also be merged with bundles of the same time. The stack not only tracks what bundleGroup we're in, but also the last "parent" bundle, or `referencingBundleRoot`. We draw an edge from the `referencingBundleRoot` to our new bundle. We preemptively attempt to merge this bundle into any existing bundle of the same type **within our bundleGroup**. Merging is also handled in the next step, we track what bundles are eligible for merging via `typeChangeIds`.
```
Comment (devongovett) [inline docs/DefaultBundler.md]:
```suggestion
Bundles created by this path are **not** bundleGroups, and can also be merged with bundles of the same type. The stack not only tracks what bundleGroup we're in, but also the last "parent" bundle, or `referencingBundleRoot`. We draw an edge from the `referencingBundleRoot` to our new bundle. We preemptively attempt to merge this bundle into any existing bundle of the same type **within our bundleGroup**. Merging is also handled in the next step, we track what bundles are eligible for merging via `typeChangeIds`.
```

File: docs/DefaultBundler.md [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+
+## Step: Merge Type Change Bundles
+
+Type change bundles are a special case of bundle, because they require consistent or “stable” names. As a result, we only allow one bundle of another type per bundleGroup. So, we need to merge bundles that exist within the same bundleGroups. (i.e. siblings)
```
Comment (devongovett) [inline docs/DefaultBundler.md]: they only require stable names in entry bundle groups right?

File: docs/DefaultBundler.md [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+
+## Step: Insert or Share
+
+Here, we finally place assets into bundles which require them. All of the previous work we've done assumes all synchronous assets will be available to the bundle that needs it, but we haven't actually placed those assets in yet. If an asset is “reachable” from many bundles, we can extract it into a shared bundle.
+
+You may think of reachable as all the bundles that still need an asset by some means. We filter it down to ensure we only place assets where they need to be. Any bundle that contains the asset in question in its ancestorAssets is filtered out, & entries are filtered out.
+
+### Reused Bundles
+
+There’s a special case here that is unique to the experimental bundler, which is reusing bundles. We noticed sometimes, if two or more bundles shared the whole contents of another bundle, reusing that bundle is as simple as drawing an edge, as opposed to creating a shared bundle that would essentially be a copy of another.
```
Comment (devongovett) [inline docs/DefaultBundler.md]: not experimental anymore... there's a few other references to it in these docs as well

File: docs/BundlerExamples.md [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+
+Here I will explain line by line how we actually place assets in the case of reused bundles.
+
+During placement, we go through each asset, one by one, and determine the set of bundles it must be placed in.
+
+```
+for (let i = 0; i < assets.length; i++) { ... }
+```
+
+Then we handle placement for entries and manual shared assets, see (DefaultBundler.md)[DefaultBundler.md] for a more detailed look at that section.
```
Comment (devongovett) [inline docs/BundlerExamples.md]:
```suggestion
Then we handle placement for entries and manual shared assets, see [DefaultBundler.md](DefaultBundler.md) for a more detailed look at that section.
```

**Review by devongovett — APPROVED** [review]
Thanks for writing these!

File: docs/DefaultBundler.md (hunk 2 above)
Comment (AGawrys) [reply docs/DefaultBundler.md]: Ahh yes, you know what, I believe we used to have it such that it merged **any** bundles of the same type within a bundleGroup...but I see now we do `bundleGraph.nodes.entries` for the groups that are up for merging. Nice catch.

Updated it to reflect this

**Pull request merged**